rfcs/function-and-table-types.md

```diff
@@ -0,0 +1,68 @@
+# Add primitive function and table types
+
+## Summary
+
+Add types for "real" functions and tables.
+
+## Motivation
+
+Some APIs require "real" functions and tables, not just things that
+"look functiony" (e.g. tables with a `__call` metamethod) or "look
+tably" (e.g. instances of classes). This RFC adds types for those.
+
+For example, the function:
+
+```lua
+  function succ(x)
+    if type(x) == "function"
+      return x() + 1
+    else
+      assert(type(x) == number)
+      return x + 1
+    end
+  end
+```
+
+cannot quite be given an accurate Luau type. The nearest is `(number |
+()->number) -> number` but this is slightly too generous, since it
+allows "functiony" types such as tables with a `__call` metamethod:
+
+```lua
+  local t = setmetatable({}, {__call = function(self) return 5 end})
+  succ(t)
+```
+
+This will typecheck but produce a runtime error.
```

Comment:
This is not true, you get an error: `Type 't' could not be converted into '(() -> number) | number'; none of the union options are compatible`
So since callable tables are not subtypes of current function types already, is the motivation for this proposal incorrect?

Reply:
Oh sigh, good point, this coercion is done at call sites, not by subtyping. The coercion from C++ functions to functions is done by subtyping, so we have an inconsistency.  I guess the question is do we think function application should use a train mechanism similar to the idea for overloaded operators, or as now with function types? That is, what do we think should happen with
```lua
 local t = setmetatable({}, {__call = function(self) return 5 end})
 function f(g) return g() end
 f(t)
```
Ot seems like there's three routes:
* type error (as now)
* OK because the type of `f` is `<a...>(() -> a...) -> a...`
* OK because the type of `f` is `<a..., b>(b) -> a... where Callable<b, (), a...>`

Reply:
I'm now leaning towards

> OK because the type of `f` is `<a..., b>(b) -> a... where Callable<b, (), a...>`

because it's just like traits for operators, `()` is essentially an n-ary operator on the function and so it's consistent with `+` and friends. Only downside with this approach is the length of types we infer, but that's already true if we go with that same approach for operators.

btw, unrelated but `Callable` should just take one type argument, not two type pack arguments. This is so that generic type parameters in functions share the scopes across the arguments and returns, `Callable<<a>(a) -> a>`. You can't really do this with `Callable<(a), (a)>`. So in this case, `b: Callable<() -> a...>`.

Reply:
I'd prefer to infer `<a...>(() -> a...) -> a...` as now. These types are user-visible, and I'd like to keep them as simple as we can.